pkg/handlers/notification.go

```diff
@@ -51,112 +64,202 @@ func getFilter(filters []string) map[string]notification.Type {
-			if err != nil {
-				middleware.GetLogger(r).Error().Msg("encoding notification to json")
-			} else {
-				messageChan <- jsonEvent
+	for notificationType := range notificationsToAdd {
+		id, e := nc.AddHandler(notificationType, func(n interface{}) {
+			msg := syncer.Event{
+				Type:    notificationType,
+				Message: n,
 			}
+			messageChan <- msg
 		})
 		if e != nil {
-			RenderError(e, http.StatusUnprocessableEntity, w, r)
-			return
+			return nil, e
 		}
 
 		// do defer outside the loop.
 		ids = append(ids, struct {
 			int
 			notification.Type
-		}{id, value})
+		}{id, notificationType})
 	}
 
-	// Remove the decision listener if we exited.
-	defer func() {
-		for _, id := range ids {
-			err := nc.RemoveHandler(id.int, id.Type)
-			if err != nil {
-				middleware.GetLogger(r).Error().AnErr("removing notification", err)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				for _, id := range ids {
+					err := nc.RemoveHandler(id.int, id.Type)
+					if err != nil {
+						logger.Err(err).AnErr("removing notification", err)
+					}
+				}
+				return
 			}
 		}
 	}()
 
-	// "raw" query string option
-	// If provided, send raw JSON lines instead of SSE-compliant strings.
-	raw := len(r.Form["raw"]) > 0
-
-	// Listen to connection close and un-register messageChan
-	notify := r.Context().Done()
-	// block waiting or messages broadcast on this connection's messageChan
-	for {
-		select {
-		// Write to the ResponseWriter
-		case msg := <-messageChan:
-			if raw {
-				// Raw JSON events, one per line
-				_, _ = fmt.Fprintf(w, "%s\n", msg)
-			} else {
-				// Server Sent Events compatible
-				_, _ = fmt.Fprintf(w, "data: %s\n\n", msg)
-			}
-			// Flush the data immediately instead of buffering it for later.
-			// The flush will fail if the connection is closed.  That will cause the handler to exit.
-			flusher.Flush()
-		case <-notify:
-			middleware.GetLogger(r).Debug().Msg("received close on the request.  So, we are shutting down this handler")
-			return
+	return messageChan, nil
+}
+
+func RedisNotificationReceiver(conf config.SyncConfig) NotificationReceiverFunc {
+	return func(ctx context.Context) (<-chan syncer.Event, error) {
+		redisSyncer, err := syncer.NewRedisNotificationSyncer(&zerolog.Logger{}, conf)
+		if err != nil {
+			return nil, err
 		}
-	}
 
+		client := redis.NewClient(&redis.Options{
+			Addr:     redisSyncer.Host,
+			Password: redisSyncer.Password,
+			DB:       redisSyncer.Database,
+		})
+
+		// Subscribe to a Redis channel
+		pubsub := client.Subscribe(ctx, redisSyncer.Channel)
+
+		dataChan := make(chan syncer.Event)
+
+		logger, ok := ctx.Value(LoggerKey).(*zerolog.Logger)
+		if !ok {
+			logger = &zerolog.Logger{}
+		}
+
+		go func() {
+			for {
+				select {
+				case <-ctx.Done():
+					client.Close()
+					pubsub.Close()
+					logger.Debug().Msg("context canceled, redis notification receiver is closed")
+					return
+				default:
+					msg, err := pubsub.ReceiveMessage(ctx)
```

Comment:
It looks like we're getting notifications for all sdkKeys. If so, we need to filter out those notifcations for the requested sdkKey only. Can we consider a channel per sdkKey?

Reply:
Yes! we can maintain a pub-sub channel for each sdk key.

Reply:
added sdk key based filtering